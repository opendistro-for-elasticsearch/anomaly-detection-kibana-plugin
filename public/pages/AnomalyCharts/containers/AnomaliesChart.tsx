/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React, { useState, useEffect } from 'react';
import ContentPanel from '../../../components/ContentPanel/ContentPanel';
import {
  EuiFlexItem,
  EuiFlexGroup,
  EuiSelect,
  EuiIcon,
  EuiLoadingChart,
  EuiStat,
} from '@elastic/eui';
import moment, { Moment } from 'moment';
import {
  Chart,
  Axis,
  LineSeries,
  niceTimeFormatter,
  Settings,
  Position,
  LineAnnotation,
  AnnotationDomainTypes,
  RectAnnotation,
} from '@elastic/charts';
import { useDelayedLoader } from '../../../hooks/useDelayedLoader';
import {
  AnomalySummary,
  Monitor,
  Detector,
  DateRange,
} from '../../../models/interfaces';
import { get } from 'lodash';
import { prepareDataForChart } from '../../utils/anomalyResultUtils';
import { AlertsFlyout } from '../components/AlertsFlyout/AlertsFlyout';

import { AlertsButton } from '../components/AlertsButton/AlertsButton';
import { darkModeEnabled } from '../../../utils/kibanaUtils';
import {
  AlertsStat,
  AnomalyStatWithTooltip,
} from '../components/AnomaliesStat/AnomalyStat';
import { AD_RESULT_DATE_RANGES } from '../../utils/constants';
import {
  getAlerts,
  convertAlerts,
  generateAlertAnnotations,
  getAnomalySummary,
  disabledHistoryAnnotations,
  INITIAL_ANOMALY_SUMMARY,
  ANOMALY_DATE_RANGE_OPTIONS,
} from '../utils/anomalyChartUtils';

interface AnomaliesChartProps {
  onDateRangeChange(
    startDate: Moment,
    endDate: Moment,
    dateRangeOption?: string
  ): void;
  title: string;
  anomalies: any[];
  annotations?: any[];
  isLoading: boolean;
  dateRange: DateRange;
  showAlerts?: boolean;
  anomalyGradeSeriesName: string;
  confidenceSeriesName: string;
  initialDateRangeOption: string;
  dateRangeOptions?: any[];
  detectorId: string;
  detectorName: string;
  detector?: Detector;
  detectorInterval?: number;
  unit?: string;
  noFeature?: boolean;
  monitor?: Monitor;
}
export const AnomaliesChart = React.memo((props: AnomaliesChartProps) => {
  const [anomalySummary, setAnomalySummary] = useState<AnomalySummary>(
    INITIAL_ANOMALY_SUMMARY
  );
  const [showAlertsFlyout, setShowAlertsFlyout] = useState<boolean>(false);
  const [totalAlerts, setTotalAlerts] = useState<number | undefined>(undefined);
  const [alertAnnotations, setAlertAnnotations] = useState<any[]>([]);
  const [isLoadingAlerts, setIsLoadingAlerts] = useState<boolean>(false);

  useEffect(() => {
    setAnomalySummary(getAnomalySummary(props.anomalies));
  }, [props.anomalies]);

  useEffect(() => {
    async function getMonitorAlerts(monitorId: string, startDateTime: number) {
      try {
        setIsLoadingAlerts(true);
        const result = await getAlerts(monitorId, startDateTime);
        setIsLoadingAlerts(false);
        setTotalAlerts(
          get(result, 'data.response.aggregations.total_alerts.value')
        );
        const monitorAlerts = convertAlerts(result);
        const annotations = generateAlertAnnotations(monitorAlerts);
        setAlertAnnotations(annotations);
      } catch (err) {
        console.error(`Failed to get alerts for monitor ${monitorId}`, err);
        setIsLoadingAlerts(false);
      }
    }
    if (props.monitor && props.dateRange.startDate) {
      getMonitorAlerts(props.monitor.id, props.dateRange.startDate.valueOf());
    }
  }, [props.monitor, props.dateRange.startDate]);

  const anomalyChartTimeFormatter = niceTimeFormatter([
    props.dateRange.startDate.valueOf(),
    props.dateRange.endDate.valueOf(),
  ]);

  const showLoader = useDelayedLoader(props.isLoading || isLoadingAlerts);

  const datePicker = () => (
    <EuiSelect
      id="anomalyHistoryDatePicker"
      options={
        props.dateRangeOptions
          ? props.dateRangeOptions
          : ANOMALY_DATE_RANGE_OPTIONS
      }
      value={props.initialDateRangeOption}
      onChange={(e: any) => {
        if (e.target.value === AD_RESULT_DATE_RANGES.LAST_1_HOUR) {
          props.onDateRangeChange(
            moment().subtract(1, 'hours'),
            moment(),
            AD_RESULT_DATE_RANGES.LAST_1_HOUR
          );
        }
        if (e.target.value === AD_RESULT_DATE_RANGES.LAST_24_HOURS) {
          props.onDateRangeChange(
            moment().subtract(24, 'hours'),
            moment(),
            AD_RESULT_DATE_RANGES.LAST_24_HOURS
          );
        }
        if (e.target.value === AD_RESULT_DATE_RANGES.LAST_7_DAYS) {
          props.onDateRangeChange(
            moment().subtract(7, 'days'),
            moment(),
            AD_RESULT_DATE_RANGES.LAST_7_DAYS
          );
        }
      }}
      style={{ width: '145px' }}
    />
  );

  const setUpAlertsButton = () => (
    <AlertsButton
      monitor={props.monitor}
      detectorId={props.detectorId}
      detectorName={props.detectorName}
      detectorInterval={get(props, 'detectorInterval', 1)}
      unit={get(props, 'unit', 'Minutes')}
    />
  );

  const anomalies = prepareDataForChart(
    props.anomalies,
    props.dateRange.startDate,
    props.dateRange.endDate
  );

  return (
    <React.Fragment>
      <ContentPanel
        title={props.title}
        actions={
          props.showAlerts
            ? [datePicker(), setUpAlertsButton()]
            : [datePicker()]
        }
      >
        <EuiFlexGroup direction="column">
          <EuiFlexGroup style={{ padding: '20px' }}>
            <EuiFlexItem>
              <EuiStat
                title={
                  props.isLoading || isLoadingAlerts
                    ? '-'
                    : anomalySummary.anomalyOccurrence
                }
                description={
                  props.showAlerts
                    ? 'Anomaly occurrences'
                    : 'Sample anomaly occurrences'
                }
                titleSize="m"
              />
            </EuiFlexItem>
            <EuiFlexItem>
              <AnomalyStatWithTooltip
                isLoading={props.isLoading || isLoadingAlerts}
                // isLoadingAlerts={isLoadingAlerts}
                minValue={anomalySummary.minAnomalyGrade}
                maxValue={anomalySummary.maxAnomalyGrade}
                description={
                  props.showAlerts ? 'Anomaly grade' : 'Sample anomaly grade'
                }
                tooltip="Indicates to what extent this data point is anomalous."
              />
            </EuiFlexItem>
            <EuiFlexItem>
              <AnomalyStatWithTooltip
                isLoading={props.isLoading || isLoadingAlerts}
                // isLoadingAlerts={isLoadingAlerts}
                minValue={anomalySummary.minConfidence}
                maxValue={anomalySummary.maxConfidence}
                description={
                  props.showAlerts ? 'Confidence' : 'Sample confidence'
                }
                tooltip="Indicates the level of confidence in the anomaly result."
              />
            </EuiFlexItem>
            <EuiFlexItem>
              <EuiStat
                title={
                  props.isLoading || isLoadingAlerts
                    ? ''
                    : anomalySummary.lastAnomalyOccurrence
                }
                description={
                  props.showAlerts
                    ? 'Last anomaly occurrence'
                    : 'Last sample anomaly occurrence'
                }
                titleSize="m"
              />
            </EuiFlexItem>
            {props.showAlerts ? (
              <EuiFlexItem>
                <AlertsStat
                  monitor={props.monitor}
                  showAlertsFlyout={() => setShowAlertsFlyout(true)}
                  totalAlerts={totalAlerts}
                  isLoading={props.isLoading}
                />
              </EuiFlexItem>
            ) : null}
          </EuiFlexGroup>
          <EuiFlexGroup>
            <EuiFlexItem grow={true}>
              <div
                style={{
                  height: '200px',
                  width: '100%',
                  opacity: showLoader ? 0.2 : 1,
                }}
              >
                {props.isLoading || isLoadingAlerts ? (
                  <EuiFlexGroup
                    justifyContent="spaceAround"
                    style={{ paddingTop: '150px' }}
                  >
                    <EuiFlexItem grow={false}>
                      <EuiLoadingChart size="xl" mono />
                    </EuiFlexItem>
                  </EuiFlexGroup>
                ) : (
                  <Chart>
                    <Settings showLegend legendPosition={Position.Right} />
                    <RectAnnotation
                      dataValues={disabledHistoryAnnotations(
                        props.dateRange,
                        props.detector
                      )}
                      id="anomalyAnnotations"
                      style={{
                        stroke: darkModeEnabled() ? 'red' : '#D5DBDB',
                        strokeWidth: 1,
                        opacity: 0.8,
                        fill: darkModeEnabled() ? 'red' : '#D5DBDB',
                      }}
                    />
                    {alertAnnotations ? (
                      <LineAnnotation
                        id="alertAnnotation"
                        domainType={AnnotationDomainTypes.XDomain}
                        dataValues={alertAnnotations}
                        marker={<EuiIcon type="bell" />}
                      />
                    ) : null}
                    <Axis
                      id="bottom"
                      position="bottom"
                      tickFormat={anomalyChartTimeFormatter}
                    />
                    <Axis
                      id="left"
                      title={'Anomaly grade / confidence'}
                      position="left"
                      domain={{ min: 0, max: 1 }}
                      showGridLines
                    />
                    <LineSeries
                      id="confidence"
                      name={props.confidenceSeriesName}
                      xScaleType="time"
                      yScaleType="linear"
                      xAccessor={'plotTime'}
                      yAccessors={['confidence']}
                      color={['#017F75']}
                      data={anomalies}
                    />
                    <LineSeries
                      id="anomalyGrade"
                      name={props.anomalyGradeSeriesName}
                      data={anomalies}
                      xScaleType="time"
                      yScaleType="linear"
                      xAccessor={'plotTime'}
                      yAccessors={['anomalyGrade']}
                      color={['#D13212']}
                    />
                  </Chart>
                )}
              </div>
            </EuiFlexItem>
          </EuiFlexGroup>
        </EuiFlexGroup>
      </ContentPanel>

      {showAlertsFlyout ? (
        <AlertsFlyout
          // @ts-ignore
          detectorId={props.detectorId}
          // @ts-ignore
          detectorName={props.detectorName}
          detectorInterval={get(props, 'detectorInterval', 1)}
          unit={get(props, 'unit', 'Minutes')}
          monitor={props.monitor}
          onClose={() => setShowAlertsFlyout(false)}
        />
      ) : null}
    </React.Fragment>
  );
});
