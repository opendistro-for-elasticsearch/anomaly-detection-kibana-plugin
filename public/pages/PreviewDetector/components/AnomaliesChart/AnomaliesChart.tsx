/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import ContentPanel from '../../../../components/ContentPanel/ContentPanel';
import { EuiFlexItem, EuiFlexGroup, EuiTitle, EuiSpacer } from '@elastic/eui';
import { DateRangePicker } from '../DateRangePicker';
import { Moment } from 'moment';
import {
  Chart,
  Axis,
  LineSeries,
  niceTimeFormatter,
  Settings,
  Position,
  RectAnnotation,
} from '@elastic/charts';
import { useDelayedLoader } from '../../../../hooks/useDelayedLoader';
import { darkModeEnabled } from '../../../../utils/kibanaUtils';

interface AnomaliesChartProps {
  onDateRangeChange(startDate: Moment, endDate: Moment): void;
  anomalies: any[];
  annotations: any[];
  isLoading: boolean;
  startDateTime: Moment;
  endDateTime: Moment;
}
export const AnomaliesChart = React.memo((props: AnomaliesChartProps) => {
  const timeFormatter = niceTimeFormatter([
    props.startDateTime.valueOf(),
    props.endDateTime.valueOf(),
  ]);
  const showLoader = useDelayedLoader(props.isLoading);
  return (
    <React.Fragment>
      <EuiFlexGroup alignItems="flexEnd">
        <EuiFlexItem>
          <EuiTitle size="s" className="preview-title">
            <h4>Detector output</h4>
          </EuiTitle>
        </EuiFlexItem>
        <EuiFlexItem>
          <DateRangePicker
            onRangeChange={props.onDateRangeChange}
            initialStartTime={props.startDateTime}
            initialEndTime={props.endDateTime}
          />
        </EuiFlexItem>
      </EuiFlexGroup>
      <EuiSpacer size="s" />
      <ContentPanel
        title="Total anomalies"
      >
        <div
          style={{
            height: '300px',
            width: '100%',
            opacity: showLoader ? 0.2 : 1,
          }}
        >
          <Chart>
            <Settings
              showLegend
              legendPosition={Position.Bottom}
              // showLegendDisplayValue={false}
              // showLegendExtra={false}
            />
            <RectAnnotation
              dataValues={props.annotations || []}
              id="annotations"
              // annotationId={getAnnotationId('react')}
              style={{
                stroke: darkModeEnabled() ? 'red' : '#FCAAAA',
                strokeWidth: 1,
                opacity: 0.8,
                fill: darkModeEnabled() ? 'red' : '#FCAAAA',
              }}
            />
            <Axis id="bottom" position="bottom" tickFormat={timeFormatter} />
            <Axis
              id="left"
              title={'Anomaly grade / confidence'}
              position="left"
              domain={{ min: 0, max: 1 }}
            />
            <LineSeries
              id="Anomaly grade"
              xScaleType="time"
              yScaleType="linear"
              xAccessor={'plotTime'}
              yAccessors={['anomalyGrade']}
              data={props.anomalies}
            />
            <LineSeries
              id="Confidence"
              xScaleType="time"
              yScaleType="linear"
              xAccessor={'plotTime'}
              yAccessors={['confidence']}
              data={props.anomalies}
            />
          </Chart>
        </div>
      </ContentPanel>
    </React.Fragment>
  );
});
