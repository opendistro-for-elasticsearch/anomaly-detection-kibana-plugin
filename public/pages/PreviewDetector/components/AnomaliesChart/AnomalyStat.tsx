/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React, { Fragment } from 'react';
import { EuiIcon, EuiLink, EuiToolTip } from '@elastic/eui';
import { Monitor } from '../../../../models/interfaces';
import { getAlertingMonitorListLink } from '../../../../utils/utils';

interface AnomalyStatProps {
  title: any;
  description: any;
}
export const AnomalyStat = (props: AnomalyStatProps) => {
  return (
    <div className="euiStat euiStat--leftAligned">
      <div className="euiText euiText--small euiStat__description">
        {props.description}
      </div>
      <div
        className="euiTitle euiTitle--medium euiStat__title"
        style={{ display: 'inline' }}
      >
        {props.title}
      </div>
    </div>
  );
};

const formatNumber = (data: any) => {
  try {
    const value = parseFloat(data);
    return !value ? value : value.toFixed(2);
  } catch (err) {
    return '';
  }
};

export const AnomalyStatWithTooltip = (props: {
  isLoading: boolean;
  isLoadingAlerts: boolean;
  minValue: number | undefined;
  maxValue: number | undefined;
  description: string;
  tooltip: string;
}) => {
  const title = () => {
    return props.isLoading || props.isLoadingAlerts
      ? '-'
      : !props.maxValue
      ? '0'
      : `${formatNumber(props.minValue)}-${formatNumber(props.maxValue)}`;
  };
  const description = () => {
    return (
      <Fragment>
        <p>
          {props.description}{' '}
          <EuiToolTip position="top" content={props.tooltip}>
            <EuiIcon type="iInCircle" />
          </EuiToolTip>
        </p>
      </Fragment>
    );
  };
  return <AnomalyStat description={description()} title={title()} />;
};

export const AlertsStat = (props: {
  monitor: Monitor | undefined;
  showAlertsFlyout(): void;
  totalAlerts: number | undefined;
  isLoading: boolean;
}) => {
  const title = () => {
    return (
      <Fragment>
        <p
          className="euiTitle euiTitle--medium euiStat__title"
          style={{ display: 'inline' }}
        >
          {props.totalAlerts === undefined || props.isLoading
            ? '-'
            : props.totalAlerts}{' '}
        </p>
        {props.monitor ? (
          <EuiLink
            href={`${getAlertingMonitorListLink()}/${
              // @ts-ignore
              props.monitor.id
            }`}
            target="_blank"
            style={{ fontSize: '15px' }}
          >
            view monitor <EuiIcon type="popout"></EuiIcon>
          </EuiLink>
        ) : null}
      </Fragment>
    );
  };

  const description = () => {
    return (
      <p>
        Alert{' '}
        <EuiLink onClick={props.showAlertsFlyout} style={{ fontSize: '15px' }}>
          info
        </EuiLink>
      </p>
    );
  };

  return <AnomalyStat title={title()} description={description()} />;
};
