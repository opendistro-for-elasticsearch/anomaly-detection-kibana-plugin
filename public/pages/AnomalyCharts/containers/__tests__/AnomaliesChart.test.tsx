/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import { render } from '@testing-library/react';
import { AnomaliesChart } from '../AnomaliesChart';
import moment from 'moment';

describe('<AnomaliesChart /> spec', () => {
  const initialStartTime = moment('2019-10-10T09:00:00');
  const initialEndTime = initialStartTime.clone().add(2, 'd');
  const dateRange = {
    startDate: initialStartTime,
    endDate: initialEndTime,
  };
  test('renders the component', () => {
    const { container } = render(
      <AnomaliesChart
        title="test"
        detectorId="testDetectorId"
        detectorName="testDetectorName"
        anomalyGradeSeriesName="anoaly grade"
        confidenceSeriesName="confidence"
        dateRangeOption="last_24_hours"
        onDateRangeChange={jest.fn()}
        anomalies={[]}
        annotations={[]}
        dateRange={dateRange}
        isLoading={false}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
