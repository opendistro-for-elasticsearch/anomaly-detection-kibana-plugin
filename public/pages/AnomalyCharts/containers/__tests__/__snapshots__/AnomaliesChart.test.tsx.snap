// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AnomaliesChart /> spec renders the component 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium"
  style="padding-left: 0px; padding-right: 0px;"
>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
    style="padding: 0px 10px;"
  >
    <div
      class="euiFlexItem"
    >
      <h3
        class="euiTitle euiTitle--small content-panel-title"
      >
        test
      </h3>
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
      >
        <div
          class="euiFlexItem"
        />
      </div>
    </div>
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      >
        <div
          class="euiFlexItem"
        >
          <div
            class="euiFormControlLayout"
          >
            <div
              class="euiFormControlLayout__childrenWrapper"
            >
              <select
                class="euiSelect"
                id="anomalyHistoryDatePicker"
                style="width: 145px;"
              >
                <option
                  value="last_1_hour"
                >
                  Last 1 hour
                </option>
                <option
                  value="last_24_hours"
                >
                  Last 24 hours
                </option>
                <option
                  value="last_7_days"
                >
                  Last 7 days
                </option>
              </select>
              <div
                class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
              >
                <span
                  class="euiFormControlLayoutCustomIcon"
                >
                  <svg
                    aria-hidden="true"
                    class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                    focusable="false"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  />
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <hr
    class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
  />
  <div
    style="padding: 0px 10px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionColumn euiFlexGroup--responsive"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        style="padding: 20px;"
      >
        <div
          class="euiFlexItem"
        >
          <div
            class="euiStat euiStat--leftAligned"
          >
            <div
              class="euiText euiText--small euiStat__description"
            >
              <p
                id="random_id"
              >
                Sample anomaly occurrences
              </p>
            </div>
            <p
              aria-labelledby="random_id"
              class="euiTitle euiTitle--medium euiStat__title"
            >
              0
            </p>
          </div>
        </div>
        <div
          class="euiFlexItem"
        >
          <div
            class="euiStat euiStat--leftAligned"
          >
            <div
              class="euiText euiText--small euiStat__description"
            >
              <p>
                Sample anomaly grade
                 
                <span
                  class="euiToolTipAnchor"
                >
                  <svg
                    class="euiIcon euiIcon--medium euiIcon-isLoading"
                    focusable="false"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  />
                </span>
              </p>
            </div>
            <div
              class="euiTitle euiTitle--medium euiStat__title"
              style="display: inline;"
            >
              0
            </div>
          </div>
        </div>
        <div
          class="euiFlexItem"
        >
          <div
            class="euiStat euiStat--leftAligned"
          >
            <div
              class="euiText euiText--small euiStat__description"
            >
              <p>
                Sample confidence
                 
                <span
                  class="euiToolTipAnchor"
                >
                  <svg
                    class="euiIcon euiIcon--medium euiIcon-isLoading"
                    focusable="false"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  />
                </span>
              </p>
            </div>
            <div
              class="euiTitle euiTitle--medium euiStat__title"
              style="display: inline;"
            >
              0
            </div>
          </div>
        </div>
        <div
          class="euiFlexItem"
        >
          <div
            class="euiStat euiStat--leftAligned"
          >
            <div
              class="euiText euiText--small euiStat__description"
            >
              <p
                id="random_id"
              >
                Last sample anomaly occurrence
              </p>
            </div>
            <p
              aria-labelledby="random_id"
              class="euiTitle euiTitle--medium euiStat__title"
            >
              -
            </p>
          </div>
        </div>
      </div>
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
      >
        <div
          class="euiFlexItem"
        >
          <div
            style="height: 200px; width: 100%; opacity: 1;"
          >
            <div
              class="echChart"
            >
              <div
                class="echChartStatus"
                data-ech-render-complete="true"
                data-ech-render-count="1"
              />
              <div
                class="echChartResizer"
              />
              <div
                style="position: absolute; left: -9999px; opacity: 0;"
              >
                <svg
                  class="euiIcon euiIcon--medium euiIcon-isLoading"
                  focusable="false"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                />
              </div>
              <div
                class="echContainer"
              >
                <div
                  class="echChartPointerContainer"
                  style="cursor: default;"
                >
                  <div
                    class="echCrosshair"
                  />
                  <svg
                    class="echHighlighter"
                  >
                    <defs>
                      <clippath
                        id="echHighlighterClipPath__843283f0-71f4-4050-8e37-b302a7f8b863"
                      >
                        <rect
                          height="0"
                          width="0"
                          x="0"
                          y="0"
                        />
                      </clippath>
                    </defs>
                    <g
                      transform="translate(0, 0) rotate(0)"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
